### PR TITLE
Remove C++11 dependencies in QPS test that break old compilers

### LINKED ISSUE #2412: gRPC build failure with gcc 4.4 compiler [-stc=C++0x]
Trying to build the gRPC library on MIPS 64 machine with gcc 4.4 compiler, using "-std=C++0x" and it runs into build errors [see error logs below] although it builds fine on a linux box with "-std=c++11" with gcc 4.6

/1tb/image/kaushikgoa/Default/porfidio/tools/mipscross-xlp/linux/bin/mips64-nlm-linux-c++ -B/1tb/image/kaushikgoa/Default/porfidio/tools/mipscross-xlp/linux -B/1tb/image/kaushikgoa/Default/porfidio/tools/mipscross-xlp/linux/lib -D__EXPORTED_HEADERS__ -fdiagnostics-show-option -Wno-error=cast-align -mabi=32 -march=xlp -pipe -Wall -Werror -Wno-unused -fno-unit-at-a-time -Wno-strict-aliasing -ggdb -O2 -DGOOGLE_PROTOBUF_ARCH_MIPS -std=c++0x -O2 -g -Wall -Wextra -Werror -Wno-long-long -Wno-unused-parameter -fPIC -I. -Iinclude -I/1tb/image/kaushikgoa/Default/porfidio/cmn/grpc-release-0_10_0/gens -I/1tb/image/kaushikgoa/Default/porfidio/openssl-regular/openssl-1.0.2c/include -I/1tb/image/kaushikgoa/Default/porfidio/cmn/protobuf-3.0.0-alpha-3/src -I/1tb/image/kaushikgoa/Default/porfidio/tools/zlib-1.1.4 -DNDEBUG -DINSTALL_PREFIX="/usr/local" -MMD -MF /1tb/image/kaushikgoa/Default/porfidio/cmn/grpc-release-0_10_0/objs/opt/src/cpp/client/secure_credentials.dep -c -o /1tb/image/kaushikgoa/Default/porfidio/cmn/grpc-release-0_10_0/objs/opt/src/cpp/client/secure_credentials.o src/cpp/client/secure_credentials.cc
In file included from include/grpc++/channel_interface.h:40,
                 from ./src/cpp/client/channel.h:39,
                 from src/cpp/client/secure_credentials.cc:37:
include/grpc++/impl/call.h: In member function 'void grpc::CallOpGenericRecvMessage::RecvMessage(R*)':
include/grpc++/impl/call.h:181: error: expected primary-expression before '[' token
include/grpc++/impl/call.h:181: error: expected primary-expression before '*' token
include/grpc++/impl/call.h:181: error: 'buf' was not declared in this scope
include/grpc++/impl/call.h:182: error: expected primary-expression before 'int'
include/grpc++/impl/call.h:182: error: expected ';' before '{' token
make: *** [/1tb/image/kaushikgoa/Default/porfidio/cmn/grpc-release-0_10_0/objs/opt/src/cpp/client/secure_credentials.o] Error 1
make: Leaving directory `/1tb/image/kaushikgoa/Default/porfidio/cmn/grpc-release-0_10_0'
 
Source code for include/grpc++/impl/call.h
175 class CallOpGenericRecvMessage {
176  public:
177   CallOpGenericRecvMessage() : got_message(false) {}
178
179   template <class R>
180   void RecvMessage(R* message) {
181     deserialize_ = [message](grpc_byte_buffer* buf,
182                              int max_message_size) -> Status {
183       return SerializationTraits<R>::Deserialize(buf, message,
184                                                  max_message_size);
185     };
186   }
187
188   bool got_message;
189
190  protected:

*************
Issue is with L181 and 182 of include/grpc++/impl/call.h

### LINKED ISSUE #2444: Please support RedHat 6
I am running into some problems building this library on a Centos-6.6 machine.
I realize RedHat-6 is a bit old, but I need to use it b/c of the EDA Consortium interoperability guideline:
http://www.edac.org/initiatives/committees/interoperability

AFAIK, gcc version on RedHat-6 is 4.4.7:

```
$ gcc --version
gcc (GCC) 4.4.7 20120313 (Red Hat 4.4.7-11)
Copyright (C) 2010 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
```

It would be very helpful to me if you could add this to your CI regressions.

### PR DESCRIPTION
**Pull request by vjpai**
This code makes the QPS tests work with gcc-4.4. In particular, the QPS tests no longer use lambdas, C++ randomization engines, or std::this_thread::sleep_until, and they only use std::vector if the contained class supports a copy constructor (More modern compilers allow copy or move constructors). This is now consistent with most of the gRPC C++ library.

Fixes #2444 
Fixes #2412

### CODE REVIEW

**Comment by vjpai** [comment]
Still need to clang-format this and add some helpful commits but my battery is almost dead... Will do later tonight

File: test/cpp/qps/client.h [hunk 1]
```diff
@@ -79,7 +83,7 @@ class Client {
 
   ClientStats Mark() {
     Histogram latencies;
-    std::vector<Histogram> to_merge(threads_.size());
+    Histogram to_merge[threads_.size()];  // avoid std::vector for old compilers
```
Comment (ctiller) [inline test/cpp/qps/client.h]: Variable sized arrays are not part of any standard.

Comment (vjpai) [inline test/cpp/qps/client.h]: I thought it looked unkosher but was just happy that it worked. I guess all our compilers support C99 extensions. That said, I need to back it out, obviously.

Comment (vjpai) [inline test/cpp/qps/client.h]: Ok, this is fixed in commit ad3e00c

**Comment by vjpai** [comment]
@dgquintas - assigning this back to you. PTAL. Thanks!

File: test/cpp/qps/client.h [hunk 2]
```diff
@@ -41,6 +41,8 @@
 
 #include <condition_variable>
 #include <mutex>
+#include <grpc++/config.h>
+#include <grpc++/config.h>
```
Comment (dgquintas) [inline test/cpp/qps/client.h]: duplicated line

File: test/cpp/qps/client.h [hunk 3]
```diff
@@ -79,7 +83,8 @@ class Client {
 
   ClientStats Mark() {
     Histogram latencies;
-    std::vector<Histogram> to_merge(threads_.size());
+    // avoid std::vector for old compilers
```
Comment (dgquintas) [inline test/cpp/qps/client.h]: WAT! really? no std::vector? Which c++ compiler that can be regarded as such doesn't have an STL impl with std::vector? (other than maybe not having it inside std::)

File: test/cpp/qps/client.h [hunk 4]
```diff
@@ -89,6 +94,7 @@ class Client {
       threads_[i]->EndSwap();
       latencies.Merge(&to_merge[i]);
     }
+    delete[] to_merge;
```
Comment (dgquintas) [inline test/cpp/qps/client.h]: delete[]... I have no memory of this place...

File: test/cpp/qps/client.h [hunk 5]
```diff
@@ -106,9 +112,20 @@ class Client {
 
   class ClientChannelInfo {
    public:
-    ClientChannelInfo(const grpc::string& target, const ClientConfig& config)
-        : channel_(CreateTestChannel(target, config.enable_ssl())),
-          stub_(TestService::NewStub(channel_)) {}
+    ClientChannelInfo() {}
+    ClientChannelInfo(const ClientChannelInfo& i) : channel_(), stub_() {
```
Comment (dgquintas) [inline test/cpp/qps/client.h]: nit: you don't need to explicitly call the default constructor, even though maybe the explicitness is a good thing.

File: test/cpp/qps/client.h [hunk 6]
```diff
@@ -226,13 +225,37 @@ class Client {
...
     }
 
    private:
     Thread(const Thread&);
     Thread& operator=(const Thread&);
 
+    void ThreadFunc() {
+      for (;;) {
+        // run the loop body
+        bool thread_still_ok = client_->ThreadFunc(&histogram_, idx_);
```
Comment (dgquintas) [inline test/cpp/qps/client.h]: ubernit: const

File: test/cpp/qps/client.h [hunk 7]
```diff
@@ -226,13 +225,37 @@ class Client {
...
+        // lock, see if we're done
+        std::lock_guard<std::mutex> g(mu_);
```
Comment (dgquintas) [inline test/cpp/qps/client.h]: We don't have std::vector but we have std::lock_guard?

File: test/cpp/qps/client_async.cc [hunk 8]
```diff
@@ -156,7 +156,7 @@ class AsyncClient : public Client {
       std::function<ClientRpcContext*(int, TestService::Stub*,
                                       const SimpleRequest&)> setup_ctx)
       : Client(config),
-        channel_lock_(config.client_channels()),
+        channel_lock_(new std::mutex[config.client_channels()]),
```
Comment (dgquintas) [inline test/cpp/qps/client_async.cc]: ditto as for std::lock_guard. Is std::mutex supported by the uberold compilers that didn't have std::vector?

File: test/cpp/qps/client_async.cc [hunk 9]
```diff
@@ -316,23 +317,28 @@ class AsyncClient : public Client {
...
-  std::vector<int> next_channel_;      // per thread round-robin channel ctr
-  std::vector<boolean> issue_allowed_; // may this thread attempt to issue
-  std::vector<grpc_time> next_issue_;  // when should it issue?
+  std::vector<int> next_channel_;       // per thread round-robin channel ctr
+  std::vector<boolean> issue_allowed_;  // may this thread attempt to issue
+  std::vector<grpc_time> next_issue_;   // when should it issue?
 
-  std::vector<std::mutex> channel_lock_;
+  std::mutex*
+      channel_lock_;  // a vector, but avoid std::vector for old compilers
```
Comment (dgquintas) [inline test/cpp/qps/client_async.cc]: Is this how clang-format splits the line?

File: test/cpp/qps/client_async.cc [hunk 10]
```diff
@@ -316,23 +317,28 @@ class AsyncClient : public Client {
   }
 
  private:
-  class boolean { // exists only to avoid data-race on vector<bool>
+  class boolean {  // exists only to avoid data-race on vector<bool>
```
Comment (dgquintas) [inline test/cpp/qps/client_async.cc]: It'd be interested to know the back-story on this one... I'm aware of the template specialization that happens in the STL for vector<bool>, but which data-races was it causing?

File: test/cpp/qps/client_async.cc [hunk 11]
```diff
@@ -348,15 +354,17 @@ class AsyncUnaryClient GRPC_FINAL : public AsyncClient {
   ~AsyncUnaryClient() GRPC_OVERRIDE { EndThreads(); }
 
  private:
+  static void CheckDone(grpc::Status s, SimpleResponse* response) {}
+  static std::unique_ptr<grpc::ClientAsyncResponseReader<SimpleResponse>>
```
Comment (dgquintas) [inline test/cpp/qps/client_async.cc]: old compilers may dislike the use of ">>"

File: test/cpp/qps/client_async.cc [hunk 12]
```diff
@@ -442,16 +450,19 @@ class AsyncStreamingClient GRPC_FINAL : public AsyncClient {
   ~AsyncStreamingClient() GRPC_OVERRIDE { EndThreads(); }
 
  private:
+  static void CheckDone(grpc::Status s, SimpleResponse* response) {}
+  static std::unique_ptr<
+      grpc::ClientAsyncReaderWriter<SimpleRequest, SimpleResponse>>
```
Comment (dgquintas) [inline test/cpp/qps/client_async.cc]: ditto for >>

File: test/cpp/qps/driver.cc [hunk 13]
```diff
@@ -77,16 +77,34 @@ static deque<string> get_hosts(const string& name) {
...
+struct ServerData {
+  unique_ptr<Worker::Stub> stub;
+  unique_ptr<ClientReaderWriter<ServerArgs, ServerStatus>> stream;
+};
+
+struct ClientData {
+  unique_ptr<Worker::Stub> stub;
+  unique_ptr<ClientReaderWriter<ClientArgs, ClientStatus>> stream;
+};
+}
```
Comment (dgquintas) [inline test/cpp/qps/driver.cc]: append `// namespace runsc` to closing brace of namespace

File: test/cpp/qps/driver.h [hunk 14]
```diff
@@ -41,10 +41,18 @@
...
+class ResourceUsage {
+ public:
+  ResourceUsage(double w, double u, double s)
+      : wall_time_(w), user_time_(u), system_time_(s) {}
+  double wall_time() { return wall_time_; }
+  double user_time() { return user_time_; }
+  double system_time() { return system_time_; }
+
+ private:
+  double wall_time_;
```
Comment (dgquintas) [inline test/cpp/qps/driver.h]: const. Same for the following two lines also.

File: test/cpp/qps/interarrival.h [hunk 15]
```diff
@@ -141,17 +142,16 @@ class ParetoDist GRPC_FINAL : public RandomDist {
...
   InterarrivalTimer() {}
   void init(const RandomDist& r, int threads, int entries = 1000000) {
-    qps_random_engine gen;
-    std::uniform_real_distribution<double> uniform(0.0, 1.0);
     for (int i = 0; i < entries; i++) {
-      random_table_.push_back(std::chrono::nanoseconds(
-          static_cast<int64_t>(1e9 * r(uniform(gen)))));
+      // rand is the only choice that is portable across POSIX and Windows
+      // and that supports new and old compilers
+      double uniform_0_1 = rand() / RAND_MAX;
```
Comment (dgquintas) [inline test/cpp/qps/interarrival.h]: const

File: test/cpp/qps/qps_driver.cc [hunk 16]
```diff
@@ -36,6 +36,7 @@
 
 #include <gflags/gflags.h>
 #include <grpc/support/log.h>
+#include <signal.h>
```
Comment (dgquintas) [inline test/cpp/qps/qps_driver.cc]: maybe move line to the top? (above "our" .h files)

File: test/cpp/qps/report.cc [hunk 17]
```diff
@@ -139,33 +132,25 @@ void PerfDbReporter::ReportQPSPerCore(const ScenarioResult& result) {
...
-                  [](ResourceUsage u) { return u.system_time; }) /
-      sum(result.client_resources, [](ResourceUsage u) { return u.wall_time; });
-  double client_user_time =
-      100.0 * sum(result.client_resources,
-                  [](ResourceUsage u) { return u.user_time; }) /
-      sum(result.client_resources, [](ResourceUsage u) { return u.wall_time; });
-
-  perf_db_client_.setTimes(server_system_time, server_user_time, client_system_time,
-                         client_user_time);
+  double server_system_time = 100.0 * sum(result.server_resources, SystemTime) /
```
Comment (dgquintas) [inline test/cpp/qps/report.cc]: const all the things!!!1 (ok, in this case it's not that useful, I just have a constant const fixation)

File: test/cpp/qps/server_async.cc [hunk 18]
```diff
@@ -142,6 +124,26 @@ class AsyncQpsServerTest : public Server {
...
 
  private:
+  void ThreadFunc(int rank) {
+    // Wait until work is available or we are shutting down
+    bool ok;
+    void *got_tag;
+    while (srv_cqs_[rank]->Next(&got_tag, &ok)) {
+      ServerRpcContext *ctx = detag(got_tag);
+      // The tag is a pointer to an RPC context to invoke
+      bool still_going = ctx->RunNextState(ok);
```
Comment (dgquintas) [inline test/cpp/qps/server_async.cc]: const

File: test/cpp/qps/client_async.cc (hunk 9 above)
Comment (vjpai) [inline test/cpp/qps/client_async.cc]: yup.

File: test/cpp/qps/client_async.cc (hunk 10 above)
Comment (vjpai) [inline test/cpp/qps/client_async.cc]: So, the C++ standard is, IMO, broken on this one. The standard officially says that accesses to disjoint elements of vectors are thread-safe, except for vector<bool>. Compilers optimize are allowed to optimize that and can let it be unsafe.

File: test/cpp/qps/client.h (hunk 2 above)
Comment (vjpai) [inline test/cpp/qps/client.h]: ack

File: test/cpp/qps/client.h (hunk 3 above)
Comment (vjpai) [inline test/cpp/qps/client.h]: done

File: test/cpp/qps/client.h (hunk 5 above)
Comment (vjpai) [inline test/cpp/qps/client.h]: I like calling the default, but I've omitted it to be consistent with the default ctor

File: test/cpp/qps/client.h (hunk 6 above)
Comment (vjpai) [inline test/cpp/qps/client.h]: Yup, that's an ubernit, but done anyway.

File: test/cpp/qps/client.h (hunk 7 above)
Comment (vjpai) [inline test/cpp/qps/client.h]: Don't know, will cross that bridge when we get there. Right now we had a customer need for 4.4

File: test/cpp/qps/client_async.cc (hunk 11 above)
Comment (vjpai) [inline test/cpp/qps/client_async.cc]: So, clang-format allows >> as do all of our compilers so far. I'd like to hold off on this.

File: test/cpp/qps/driver.cc (hunk 13 above)
Comment (vjpai) [inline test/cpp/qps/driver.cc]: done

File: test/cpp/qps/driver.h (hunk 14 above)
Comment (vjpai) [inline test/cpp/qps/driver.h]: done, I think?

File: test/cpp/qps/interarrival.h (hunk 15 above)
Comment (vjpai) [inline test/cpp/qps/interarrival.h]: done

File: test/cpp/qps/qps_driver.cc (hunk 16 above)
Comment (vjpai) [inline test/cpp/qps/qps_driver.cc]: done

File: test/cpp/qps/report.cc (hunk 17 above)
Comment (vjpai) [inline test/cpp/qps/report.cc]: done

File: test/cpp/qps/server_async.cc (hunk 18 above)
Comment (vjpai) [inline test/cpp/qps/server_async.cc]: done

**Comment by vjpai** [comment]
@dgquintas - You are right that lock_guard, unique_lock, and mutex do not seem to be supported by MSVC 2010. None of these are used in src/cpp (since we have already used grpc:: instead of std:: for each of those) but they are used in numerous places in test/cpp . I would suggest a different PR to address all of those. I will file an issue about this.

File: test/cpp/qps/client_async.cc (hunk 10 above)
Comment (dgquintas) [inline test/cpp/qps/client_async.cc]: Ah yes, I remember now. Thanks!

File: test/cpp/qps/client_async.cc (hunk 11 above)
Comment (dgquintas) [inline test/cpp/qps/client_async.cc]: Ok. In any case it's obvious to spot the problem if it were ever to occur with a really old compiler.

**Comment by dgquintas** [comment]
Sigh, unfortunate... ack.

**Comment by vjpai** [comment]
Track MSVC support in issue #2898

File: test/cpp/qps/client.h (hunk 4 above)
Comment (vjpai) [inline test/cpp/qps/client.h]: It's the 90s, go for it.

**Comment by vjpai** [comment]
Is this green enough?

**Comment by dgquintas** [comment]
Yes, LGTM.

**Comment by vjpai** [comment]
Can you merge it then?

**Comment by dgquintas** [comment]
lol yes, sorry. Multitasking doesn't pay kids, stay in school...

**Pull request merged**